proposals/ssss-distributed-actor-runtime.md

```diff
@@ -0,0 +1,1798 @@
+    }
+    self.int = int
+    // ~~~ become fully initialized ~~~
+    // << self.actorSystem.actorReady(self)
+  }
+  
+  // deinit {
+  //   << self.actorSystem.resignID(self.id)
+  // }
+}
+```
+
+Due to rules about actor and class init/deinit, when we `return nil` from a failable initializer, its deinitializer *does not run* (!). Because of this, we cannot rely on the deinit to resign the ID as we'd leave an un-used, but still registered identity hanging in the actor system, and the `resignID` is injected just before the "failing return" from such initializer. This is done transparently, and neither distributed actor developers nor actor system developers need to worry about this: the ID is always resigned properly.
+
+> This does mean that `resignID` may be called without `actorReady` having ever been called! The system should react to this as it would to any usual resignID and free any resources associated with the identifier.
+
+Next, we need to discuss *throwing* initializers, and their multiple paths of execution. Again, rules about class and actor deinitialization, are tightly related to whether or not a types deinit will be executed or not, so let us analyse the following example:
+
+```swift
+distributed actor DA {
+  var int: Int
+  
+  init(int: Int, system: ActorSystem) throws {
+    // << self.id = system.assignID(Self.self)
+    // ...
+    if int <= 1 {
+      // ...
+      // << self.actorSystem.resignID(self.id)
+      throw Boom() // [1]
+    }
+    
+    if int <= 2 {
+      self.int = int
+      // ~ become fully initialized ~
+      throw Boom() // [2]
+    }
+    
+    throw Boom() // Boom for good measure... (same as [2] though)
+    // theoretically, the ready call is inserted at the end of the init:
+    // << system.actorReady(self)
+  }
+  
+  init(int: Int, system: ActorSystem) async throws {
+    // << self.id = system.assignID(Self.self)
+    // ...
+    if int <= 1 {
+      // ...
+      // << self.actorSystem.resignID(self.id)
+      throw Boom() // [1]
+    }
+    
+    if int <= 2 {
+      self.int = int
+      // ~ become fully initialized ~
+      // << system.actorReady(self)
+      throw Boom() // [2]
+    }
+    
+    throw Boom() // Boom for good measure... (same as [2] though)
+  }
+  
+  // deinit {
+  //   << self.actorSystem.resignID(self.id)
+  // }
+}
+```
+
+The actor shown above both has state that it needs to initialize, and it is going to throw. It will trow either before becoming fully initialized `[1]`, or after it has initialized all of its stored properties `[2]`. Swift handles those two executions differently. Only a fully initialized reference type's `deinit` is going to be executed. This means that if the init throws at `[1]` we need to inject a `resignID` call there, while if it throws after becoming fully initialized, e.g. on line `[2]` we do not need to inject the `resignID` call, because the actor's `deinit` along with the injected-there `resignID` will be executed instead.
```

Comment:
```suggestion
The actor shown above both has state that it needs to initialize, and it is going to throw. It will throw either before becoming fully initialized `[1]`, or after it has initialized all of its stored properties `[2]`. Swift handles those two executions differently. Only a fully initialized reference type's `deinit` is going to be executed. This means that if the `init` throws at `[1]` we need to inject a `resignID` call there, while if it throws after becoming fully initialized, e.g. on line `[2]` we do not need to inject the `resignID` call, because the actor's `deinit` along with the injected-there `resignID` will be executed instead.
```